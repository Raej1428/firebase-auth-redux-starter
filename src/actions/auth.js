/* @flow */
import Firebase from 'firebase';
import type {
  authUserActionType,
  authErrorActionType,
  Dispatch,
} from '../types/auth';

require('firebase/firestore');

// pull out firebase into its own file soon.
const config = {
  apiKey: 'AIzaSyC3en9Q3XwoyhlyCY8wQbXK8MSVVBg_H6s',
  authDomain: 'nextjs-firebase-auth-test.firebaseapp.com',
  databaseURL: 'https://nextjs-firebase-auth-test.firebaseio.com',
  projectId: 'nextjs-firebase-auth-test',
  storageBucket: 'nextjs-firebase-auth-test.appspot.com',
  messagingSenderId: '90669433896',
};

Firebase.initializeApp(config);
// Initialize Cloud Firestore through Firebase
// $FlowFixMe
const db = Firebase.firestore();

export function authUser(uid: string): authUserActionType {
  return {
    type: 'AUTH_USER',
    payload: uid,
  };
}

export function authError(error: { code: string, message: string }): authErrorActionType {
  return {
    type: 'AUTH_ERROR',
    payload: error,
  };
}

export function createUserRecord(uid: string, firstName: string, lastName: string, email: string) {
  return db.collection('users').doc(uid).set({
    firstName,
    lastName,
    email,
    uid,
  });
}

export function signUpUser(firstName: string, lastName: string, email: string, password: string) {
  return (dispatch: Dispatch) =>
    Firebase.auth().createUserWithEmailAndPassword(email, password)
      .then((user) => {
        createUserRecord(user.uid, firstName, lastName, email);
        dispatch({
          type: 'RECORD_USER_DETAILS',
          payload: {
            uid: user.uid,
            firstName,
            lastName,
            email,
          },
        });
      });
}

export function signInUser(credentials: { email: string, password: string }) {
  return (dispatch: Dispatch) =>
    Firebase.auth().signInWithEmailAndPassword(credentials.email, credentials.password)
      .catch((error) => {
        dispatch(authError(error));
      });
}

export function signOutUser() {
  return (dispatch: Dispatch) => Firebase.auth().signOut()
    .then(() => {
      dispatch({
        type: 'SIGN_OUT_USER',
      });
    });
}

export function sendPasswordResetEmail(email: string) {
  return Firebase.auth().sendPasswordResetEmail(email);
}

export function verifyAuth() {
  return (dispatch: Dispatch) => {
    Firebase.auth().onAuthStateChanged((user) => {
      if (user) {
        dispatch(authUser(user.uid));
      } else {
        dispatch(signOutUser());
      }
    });
  };
}

