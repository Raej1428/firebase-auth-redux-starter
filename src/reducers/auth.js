/* @flow */
import type { State, Action } from '../types/auth';

const initialState: State = {
  authenticated: false,
  error: null,
  uid: null,
  profile: {},
};

export default function reducer(state: State = initialState, action: Action) {
  switch (action.type) {
    case 'AUTH_USER':
      return {
        ...state,
        authenticated: true,
        error: null,
        uid: action.payload,
      };
    case 'SIGN_OUT_USER':
      return {
        ...state,
        authenticated: false,
        error: null,
        uid: null,
        profile: {},
      };
    case 'AUTH_ERROR':
      return {
        ...state,
        error: action.payload.message,
        uid: null,
        profile: {},
      };
    case 'RECORD_USER_DETAILS':
      return {
        ...state,
        profile: action.payload,
      };
    default:
      return state;
  }
}
