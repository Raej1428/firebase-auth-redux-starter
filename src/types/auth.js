/* @flow */
/* eslint-disable */

export type State = {
  +authenticated: boolean,
  +error: string | null,
    +uid: string | null,
};

export type authUserActionType = {
  type: 'AUTH_USER',
  payload: string
}

export type signOutUserActionType = {
  type: 'SIGN_OUT_USER'
}

export type updateUserProfileActionType = {
  type: 'UPDATE_USER_DETAILS',
  payload: {
    firstName?: string,
    lastName?: string
  }
}

export type authErrorActionType = {
  type: 'AUTH_ERROR',
  payload: {
    code: string,
    message: string
  }
}

export type recordUserDetails = {
  type: 'RECORD_USER_DETAILS',
  payload: {
    uid: string,
    firstName: string,
    lastName: string,
    email: string,
  },
}

export type Action =
  | authUserActionType
  | authErrorActionType
  | signOutUserActionType
  | recordUserDetails
  | updateUserProfileActionType

export type GetState = () => State;
export type PromiseAction = Promise<Action>;
export type ThunkAction = (dispatch: Dispatch, getState: GetState) => any;
export type Dispatch = (action: Action | ThunkAction | PromiseAction | Array<Action>) => any;
