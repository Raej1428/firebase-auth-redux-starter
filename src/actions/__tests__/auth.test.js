/* eslint-disable */
import * as authActions from '../auth'

describe('AuthActions', () => {
  it('authUser action creator matches snapshot ', () => {
    const payload = authActions.authUser("12345")
    expect(payload).toMatchSnapshot();
  })

  it('authUser action creator matches snapshot ', () => {
    const payload = authActions.authError({ code: "12345", message: "error!" })
    expect(payload).toMatchSnapshot();
  })

  // Need to mock / test firebase specific stuff someday

})
