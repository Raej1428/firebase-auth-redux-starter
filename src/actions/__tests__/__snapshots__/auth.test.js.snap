// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthActions authUser action creator matches snapshot  1`] = `
Object {
  "payload": "12345",
  "type": "AUTH_USER",
}
`;

exports[`AuthActions authUser action creator matches snapshot  2`] = `
Object {
  "payload": Object {
    "code": "12345",
    "message": "error!",
  },
  "type": "AUTH_ERROR",
}
`;
